openquake/hazardlib/stats.py

```diff
@@ -89,14 +86,77 @@ def quantile_curve(curves, quantile, weights=None):
         return result_curve
 
 
-def mean_quantiles(curves_by_rlz, quantiles=(), weights=None):
+# NB: this is a function linear in the array argument
+def compute_stats(array, quantiles, weights):
+    """
+    :param array:
+        an array of R elements (which can be arrays)
+    :param quantile:
+        a list of Q quantiles
+    :param weights:
+        a list of R weights
+    :returns:
+        an array of Q + 1 elements (which can be arrays)
     """
-    :param curves_by_rlz: a list of R arrays of shape N x L
-    :param quantiles: a list of quantile PoEs (can be empty)
-    :param weights: a list of R weights (or None)
-    :returns: a list [mean, quantile...]
+    result = numpy.zeros((len(quantiles) + 1,) + array.shape[1:], array.dtype)
+    result[0] = mean_curve(array, weights)
+    shp = result[0].shape
+    for i, q in enumerate(quantiles, 1):
+        qc = quantile_curve(array, q, weights)
+        # TODO: try to simplify the ugliness below
+        if isinstance(qc, list) and len(qc) == 1:
+            result[i] = qc[0]
+        elif qc.shape != shp:
+            result[i] = qc.reshape(shp)
+        else:
+            result[i] = qc
+    return result
+
+
+# like compute_stats, but on a matrix of shape (N, R)
+def compute_stats2(arrayNR, quantiles, weights):
+    """
+    :param arrayNR:
+        an array of (N, R) elements
+    :param quantile:
+        a list of Q quantiles
+    :param weights:
+        a list of R weights
+    :returns:
+        an array of (N, Q + 1) elements
+    """
+    newshape = list(arrayNR.shape)
+    newshape[1] = len(quantiles) + 1  # number of statistical outputs
+    newarray = numpy.zeros(newshape, arrayNR.dtype)
+    data = [arrayNR[:, i] for i in range(len(weights))]
+    newarray[:, 0] = apply_stat(mean_curve, data, weights)
+    for i, q in enumerate(quantiles, 1):
+        newarray[:, i] = apply_stat(quantile_curve, data, q, weights)
+    return newarray
+
+
+def apply_stat(f, arraylist, *extra, **kw):
+    """
+    :param f: a callable arraylist -> array (of the same shape and dtype)
+    :param arraylist: a list of arrays of the same shape and dtype
+    :param extra: additional arguments
+    :returns: an array of the same shape and dtype
```

Comment:
I suppose, if you specify the param `extra`, you should also specify the param `kw`